feat: add jellyfin/emby quick connect authentication

fallenbagel opened the pull request:
## Description
Adds support for Jellyfin/Emby Quick Connect authentication, providing users with an alternative login method. Jellyfin Quick Connect button is handled as a modal (see screenshots below).

- Fixes #1595

## How Has This Been Tested?

- [X] Tested Quick Connect flow with Jellyfin server
- [ ] Tested Quick Connect flow with Emby server (Needs to be tested on emby)
- [x] Verified code expiration
- [X] Verified failed connections
- [X] Properly logs in the existing User

## Screenshots / Logs (if applicable)
Login Page:
<img width="2600" height="2086" alt="image" src="https://github.com/user-attachments/assets/d0f53536-2e25-4dae-a57a-73e19389b91e" />

Quick Connect Modal:
<img width="2600" height="2086" alt="image" src="https://github.com/user-attachments/assets/8cd74067-038e-480b-83d2-a0204c4a30c6" />

Linked Accounts modal in User profile:
<img width="1389" height="681" alt="image" src="https://github.com/user-attachments/assets/fcf65fed-f508-47e9-a750-01a1a7e13a38" />

Linked Accounts quick connect modal:
<img width="1386" height="682" alt="image" src="https://github.com/user-attachments/assets/54ab7b23-e4ac-4459-bb53-50192881cb68" />

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] I have read and followed the contribution [guidelines](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md).
- [ ] Disclosed any use of AI (see our [policy](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md#ai-assistance-notice))
- [X] I have updated the documentation accordingly.
- [X] All new and existing tests passed.
- [X] Successful build `pnpm build`
- [ ] Translation keys `pnpm i18n:extract`
- [ ] Database migration (if required)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Jellyfin/Emby “Quick Connect” sign-in with interactive modal, live polling, retry and expiration handling.
  * Link Jellyfin/Emby accounts via Quick Connect from user settings, with duplicate-account detection and feature gating.
  * New reusable Quick Connect hook and UI flows plus English localization for all Quick Connect interactions.

* **Tests**
  * Added test coverage for Quick Connect authentication, status polling, linking flows, and error conditions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fallenbagel committed:
feat: add jellyfin/emby quick connect authentication

Implements a quick connect authentication flow for jellyfin and emby servers.

fix #1595

fallenbagel committed:
feat: add jellyfin/emby quick connect authentication

Implements a quick connect authentication flow for jellyfin and emby servers.

fix #1595

fallenbagel committed:
chore(i18n): extracted translations

fallenbagel committed:
chore(i18n): extracted translations

fallenbagel committed:
fix(quick-connect): prevent memory leak by having one active poll at a time

fallenbagel committed:
fix(quick-connect): prevent memory leak by having one active poll at a time

fallenbagel committed:
fix(quick-connect): prevent multiple initiations of Quick Connect

fallenbagel committed:
fix(quick-connect): prevent multiple initiations of Quick Connect

v3DJG6GL commented:
wow, thanks a lot for this PR :D

I've built an docker image with your changes. Creating a new account with new _QuickConnect_ method works flawlessly if _Enable New Jellyfin Sign-In_ is enabled :)

However, QuickConnect functionality is missing from _Linked Accounts_ settings when trying to link a Jellyfin Account. And since my Jellyfin users are using OIDC-based login for Jellyfin - and they therefore don't have a password they could enter there - linking their Jellyfin accounts is not possible without QuickConnect functionality....

fallenbagel committed:
feat(linked-accounts): add quick connect linking in the linked-accounts module

fallenbagel committed:
feat(linked-accounts): add quick connect linking in the linked-accounts module

fallenbagel committed:
chore(i18n): extracted translations

fallenbagel committed:
chore(i18n): extracted translations

fallenbagel committed:
feat(openapi): add quick connect endpoint for linking jellyfin/emby accounts

fallenbagel committed:
feat(openapi): add quick connect endpoint for linking jellyfin/emby accounts

fallenbagel committed:
refactor(quickconnect): validate secret length and format in quick connect check

fallenbagel committed:
refactor(quickconnect): validate secret length and format in quick connect check

fallenbagel committed:
refactor(jellyfin-login): simplify error handling for quick connect errors

fallenbagel committed:
refactor(jellyfin-login): simplify error handling for quick connect errors

fallenbagel committed:
refactor(quickconnect): implement useQuickConnect hook for managing quick connect flow

fallenbagel committed:
refactor(quickconnect): implement useQuickConnect hook for managing quick connect flow

fallenbagel committed:
refactor(quickconnect): improve secret validation for quick connect endpoints

fallenbagel committed:
refactor(quickconnect): improve secret validation for quick connect endpoints

fallenbagel committed:
fix: fixes some typos

fallenbagel committed:
fix: fixes some typos

fallenbagel committed:
chore(i18n): update extracted translations

fallenbagel committed:
chore(i18n): update extracted translations

fallenbagel committed:
refactor: fix some linter issues

fallenbagel committed:
refactor: fix some linter issues

fallenbagel committed:
test(quickconnect): add tests for jellyfin qc endpoints

fallenbagel committed:
test(quickconnect): add tests for jellyfin qc endpoints

fallenbagel commented on seerr-api.yml:
```diff
@@ -3984,6 +3984,85 @@ paths:
+                    example: 'abc123def456'
+        '500':
+          description: Failed to initiate Quick Connect
+  /auth/jellyfin/quickconnect/check:
+    get:
+      summary: Check Quick Connect authorization status
+      description: Checks if the Quick Connect code has been authorized by the user.
+      security: []
+      tags:
+        - auth
+      parameters:
+        - in: query
+          name: secret
+          required: true
+          schema:
+            type: string
+          description: The secret returned from the initiate endpoint
+      responses:
+        '200':
+          description: Authorization status returned
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  authenticated:
+                    type: boolean
+                    example: false
+        '404':
+          description: Quick Connect session not found or expired
```
the secret is a short-lived opaque token that only checks status (read-only), expires automatically, and gets invalidated after use. It's not a credential like a password or API key. Jellyfin's own /QuickConnect/Connect uses get with the secret as a query param for the same reason. post for a status check would also be semantically wrong since you're not creating or mutating anything, and just polling for state. get should be the correct HTTP verb here

fallenbagel commented on server/routes/auth.ts:
```diff
@@ -593,6 +593,189 @@ authRoutes.post('/jellyfin', async (req, res, next) => {
+          }
+        );
+
+        user = new User({
+          email: account.User.Name,
+          jellyfinUsername: account.User.Name,
+          jellyfinUserId: account.User.Id,
+          jellyfinDeviceId: deviceId,
+          permissions: settings.main.defaultPermissions,
+          userType:
+            settings.main.mediaServerType === MediaServerType.JELLYFIN
+              ? UserType.JELLYFIN
+              : UserType.EMBY,
+        });
+        user.avatar = getUserAvatarUrl(user);
+        await userRepository.save(user);
```
This is intentional and consistent with all other Jellyfin user creation paths (regular login, user import). Jellyfin doesn't store email addresses, so every non-admin user falls back to username as the email field. The userEmailRequired warning exists to handle this as  it prompts users to set a real email in their settings if the admin has enabled email notifications. Adding an optional email param to QC wouldn't make sense either, since the whole flow is code-based authorization from the Jellyfin app with no form input.

fallenbagel committed:
fix: address qc review feedback

- Fix double onAuthenticated/onClose calls in QC modals
- Propagate upstream error status in usersettings QC linking endpoint
- Fix stale polling interval ref in useQuickConnect cleanup
- Stop polling when modal is hidden (show=false)
- Clear stale code on QC re-initiation
- Pass actual show prop in LinkJellyfinQuickConnectModal
- Document 400 response in QC linking OpenAPI spec
- Harden tests with call count assertions and explicit status checks

fallenbagel committed:
fix: address qc review feedback

- Fix double onAuthenticated/onClose calls in QC modals
- Propagate upstream error status in usersettings QC linking endpoint
- Fix stale polling interval ref in useQuickConnect cleanup
- Stop polling when modal is hidden (show=false)
- Clear stale code on QC re-initiation
- Pass actual show prop in LinkJellyfinQuickConnectModal
- Document 400 response in QC linking OpenAPI spec
- Harden tests with call count assertions and explicit status checks

fallenbagel commented on seerr-api.yml:
```diff
@@ -5033,6 +5112,38 @@ paths:
           description: Unlink request invalid
         '404':
           description: User does not exist
+  /user/{userId}/settings/linked-accounts/jellyfin/quickconnect:
+    post:
+      summary: Link Jellyfin/Emby account with Quick Connect
+      description: Links a Jellyfin/Emby account to the user's profile using Quick Connect authentication
+      tags:
+        - users
+      parameters:
+        - in: path
+          name: userId
+          required: true
+          schema:
+            type: number
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                secret:
+                  type: string
+              required:
+                - secret
+      responses:
+        '204':
+          description: Account successfully linked
+        '401':
+          description: Unauthorized
+        '422':
+          description: Account already linked
+        '500':
+          description: Server error
```
This is already fixed in commit https://github.com/seerr-team/seerr/commit/747fc53c8550dda9d2bda69fe96616c313c6a61b

fallenbagel committed:
refactor(quickconnect): implement zod validation for quick connect secret

fallenbagel committed:
refactor(quickconnect): implement zod validation for quick connect secret

fallenbagel committed:
fix(quick-connect): add initial valaues to useRef for React 19 compatibility

gauthier-th commented on server/routes/user/usersettings.ts:
```diff
@@ -514,6 +515,78 @@ userSettingsRoutes.delete<{ id: string }>(
+
+      const user = req.user;
+      const deviceId = Buffer.from(
+        `BOT_seerr_qc_link_${account.User.Id}`
```
Shouldn't we make this the same as [`/linked-accounts/jellyfin`](https://github.com/seerr-team/seerr/blob/6e9ba06e88d283c4c5c409c9fdac6d74ed8feaa1/server/routes/user/usersettings.ts#L394-L396)?
```ts
    const deviceId = Buffer.from(
      req.user?.id === 1 ? 'BOT_seerr' : `BOT_seerr_${req.user.username ?? ''}`
    ).toString('base64');
```

fallenbagel replied:
Let me update to match `/linked-accounts/jellyfin`. 

Though for future, `User.Name` can drift if the user renames in Jellyfin (same as the existing import path), so `deviceId` isn't fully stable long-term. Probably fine for now, but we should probably follow-up if it matters.

gauthier-th commented on server/routes/auth.ts:
```diff
@@ -593,6 +602,177 @@ authRoutes.post('/jellyfin', async (req, res, next) => {
+
+      const deviceId = Buffer.from(`BOT_seerr_qc_${account.User.Id}`).toString(
+        'base64'
+      );
```
Same comment here

fallenbagel replied:
Same as above

gauthier-th reviewed (commented):
Couldn't some UI part of `<JellyfinQuickConnectModal>` and `<LinkJellyfinQuickConnectModal.tsx‎>` be refactored to avoid code duplication?

Otherwise LGTM.

fallenbagel committed:
fix: align Quick Connect deviceId generation and deduplicate models

gauthier-th reviewed (approved):
LGTM
